### PR TITLE
update dataflatten tables to parse raw_data in query

### PR DESCRIPTION
**Pull request by James-Pickett**
allow for passing raw data into query for dataflatten tables

```
osquery> select * from kolide_json where raw_data = '[0]';
+---------+--------+-----+-------+-------+------+----------+
| fullkey | parent | key | value | query | path | raw_data |
+---------+--------+-----+-------+-------+------+----------+
| 0       |        | 0   | 0     | *     |      | [0]      |
+---------+--------+-----+-------+-------+------+----------+
osquery> select * from kolide_json where raw_data = '{"hi":"there"}';
+---------+--------+-----+-------+-------+------+----------------+
| fullkey | parent | key | value | query | path | raw_data       |
+---------+--------+-----+-------+-------+------+----------------+
| hi      |        | hi  | there | *     |      | {"hi":"there"} |
+---------+--------+-----+-------+-------+------+----------------+
```

when adding both constraints with an OR, the first one wins unless there is a specific query

```
osquery> select * from kolide_json where path = '/tmp/test.json' or raw_data = '[0]';
+---------+--------+-----+-------+-------+----------------+----------+
| fullkey | parent | key | value | query | path           | raw_data |
+---------+--------+-----+-------+-------+----------------+----------+
| hi      |        | hi  | there | *     | /tmp/test.json |          |
+---------+--------+-----+-------+-------+----------------+----------+
osquery> select * from kolide_json where query = 0 and path = '/tmp/test.json' or raw_data = '[0]';
+---------+--------+-----+-------+-------+------+----------+
| fullkey | parent | key | value | query | path | raw_data |
+---------+--------+-----+-------+-------+------+----------+
| 0       |        | 0   | 0     | *     |      | [0]      |
+---------+--------+-----+-------+-------+------+----------+
```

not sure if this is the expected behavior

### CODE REVIEW

File: ee/tables/dataflattentable/tables.go [hunk 1]
```diff
@@ -165,16 +173,43 @@ func (t *Table) generate(ctx context.Context, queryContext table.QueryContext) (
...
+	if err != nil {
+		t.slogger.Log(ctx, slog.LevelInfo,
+			"failure parsing raw data",
+			"err", err,
+		)
+		return nil, fmt.Errorf("parsing data: %w", err)
 	}
 
+	rowData := map[string]string{
+		"raw_data": rawdata,
```
Comment (James-Pickett) [inline ee/tables/dataflattentable/tables.go:206]: not sure if we should actually include this, can return a lot of data

I suppose you could always just select columns you want in query.

Comment (directionless) [reply ee/tables/dataflattentable/tables.go:206]: Sadly, we have to. If you don't, sqlite will filter out the rows. I'm not sure if we have another option today

File: ee/tables/dataflattentable/tables.go [hunk 2]
```diff
@@ -137,8 +138,15 @@ func (t *Table) generate(ctx context.Context, queryContext table.QueryContext) (
 	var results []map[string]string
 
 	requestedPaths := tablehelpers.GetConstraints(queryContext, "path")
-	if len(requestedPaths) == 0 {
-		return results, fmt.Errorf("The %s table requires that you specify a single constraint for path", t.tableName)
+	requestedRawDatas := tablehelpers.GetConstraints(queryContext, "raw_data")
+
+	if len(requestedPaths)+len(requestedRawDatas) == 0 {
```
Comment (directionless) [inline ee/tables/dataflattentable/tables.go:143]: I'd probably `len(requestedPaths) == 0 && len(requestedRawDatas) == 0`, but okay...

File: ee/tables/dataflattentable/tables.go [hunk 3]
```diff
@@ -137,8 +138,15 @@ func (t *Table) generate(ctx context.Context, queryContext table.QueryContext) (
...
+		return results, fmt.Errorf("The %s table requires that you specify a single constraint for path or raw_data", t.tableName)
```
Comment (directionless) [inline ee/tables/dataflattentable/tables.go:144]:
```suggestion
		return results, fmt.Errorf("The %s table requires that you specify a at least one of  `path` or `raw_data`", t.tableName)
```

**Review by directionless — COMMENTED** [review]
I think probably good enough

File: ee/tables/dataflattentable/tables.go (hunk 1 above)
Comment (directionless) [reply ee/tables/dataflattentable/tables.go:206]: Er, at least for the case where `raw_data` is provided. We should skip for file ones

File: ee/tables/dataflattentable/tables.go [hunk 4]
```diff
@@ -165,16 +173,43 @@ func (t *Table) generate(ctx context.Context, queryContext table.QueryContext) (
 			}
 		}
 	}
+
+	for _, rawdata := range requestedRawDatas {
+		for _, dataQuery := range tablehelpers.GetConstraints(queryContext, "query", tablehelpers.WithDefaults("*")) {
+			subresults, err := t.generateRawData(ctx, rawdata, dataQuery, append(flattenOpts, dataflatten.WithQuery(strings.Split(dataQuery, "/")))...)
+			if err != nil {
+				t.slogger.Log(ctx, slog.LevelInfo,
+					"failed to get data for raw_data",
```
Comment (directionless) [inline ee/tables/dataflattentable/tables.go:182]:
```suggestion
					"failed to generate for raw_data",
```

**Review by directionless — DISMISSED** [review]
I think it's good

File: ee/tables/dataflattentable/tables.go [hunk 5]
```diff
@@ -137,8 +138,15 @@ func (t *Table) generate(ctx context.Context, queryContext table.QueryContext) (
...
-		return results, fmt.Errorf("The %s table requires that you specify a single constraint for path", t.tableName)
+	requestedRawDatas := tablehelpers.GetConstraints(queryContext, "raw_data")
+
+	if len(requestedPaths) == 0 && len(requestedRawDatas) == 0 {
+		return results, fmt.Errorf("The %s table requires that you specify at least one of 'path' or 'raw_data'", t.tableName)
+	}
+
+	flattenOpts := []dataflatten.FlattenOpts{
+		dataflatten.WithSlogger(t.slogger),
+		dataflatten.WithNestedPlist(),
```
Comment (RebeccaMahany) [inline ee/tables/dataflattentable/tables.go:149]: Why this plist-related option?

File: ee/tables/dataflattentable/tables.go [hunk 6]
```diff
@@ -151,7 +159,7 @@ func (t *Table) generate(ctx context.Context, queryContext table.QueryContext) (
 
 		for _, filePath := range filePaths {
 			for _, dataQuery := range tablehelpers.GetConstraints(queryContext, "query", tablehelpers.WithDefaults("*")) {
-				subresults, err := t.generatePath(ctx, filePath, dataQuery)
+				subresults, err := t.generatePath(ctx, filePath, dataQuery, append(flattenOpts, dataflatten.WithQuery(strings.Split(dataQuery, "/")))...)
```
Comment (RebeccaMahany) [inline ee/tables/dataflattentable/tables.go:162]: Why do we split the query on forward slash here and on 179?

**Review by RebeccaMahany — COMMENTED** [review]
Just a couple questions to understand the changes better, but overall this looks good to me 🙂

File: ee/tables/dataflattentable/tables.go (hunk 5 above)
Comment (James-Pickett) [reply ee/tables/dataflattentable/tables.go:149]: This code handles plist in addition to json, xml, ini, jwt. I just copied these options from the original [generatePath](https://github.com/kolide/launcher/blob/main/ee/tables/dataflattentable/tables.go#L174) function to consolidate them here for both `generatePath` and `generateRawData`

File: ee/tables/dataflattentable/tables.go (hunk 6 above)
Comment (James-Pickett) [reply ee/tables/dataflattentable/tables.go:162]: I'm not quite sure, I just copied it over form the [original code](https://github.com/kolide/launcher/blob/main/ee/tables/dataflattentable/tables.go#L175), I'll look into it a little more.

Comment (James-Pickett) [reply ee/tables/dataflattentable/tables.go:162]: Hmm searching the code for `dataflatten.WithQuery(strings.Split(dataQuery, "/")` reveals that we do this in many (all?) dataflatten tables, we should probably just bake that directly into dataflatten.

**Review by RebeccaMahany — APPROVED** [review]

File: ee/tables/dataflattentable/tables.go (hunk 5 above)
Comment (directionless) [reply ee/tables/dataflattentable/tables.go:149]: `dataflatten.WithNestedPlist`is an older attempt at handling this nesting. We should carry it through

**Review by zackattack01 — APPROVED** [review]

**Pull request merged**